presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+                QualifiedName count = QualifiedName.of("count");
+                Symbol symbol = entry.getKey();
+                if (call.getName().equals(count)) {
+                    aggregations.put(symbol, new FunctionCall(
+                            count,
+                            ImmutableList.of(nonNullableAggregationSourceSymbol.toSymbolReference())));
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
+        {
+            Optional<FilterNode> filter = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .findFirst();
+
+            if (!filter.isPresent()) {
+                return Optional.of(new RewrittenScalarAggregationSource(ImmutableList.of(), aggregationSource));
+            }
+
+            Expression predicate = filter.get().getPredicate();
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
+            ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
+            for (Expression conjunct : extractConjuncts(predicate)) {
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
```

Comment:
Using `DependencyExtractor.extractUnique`, which returns a `Set`, may be more efficient, since the call `contains` below will be O(1) vs O(N) on the number of symbols in the predicate. This kind of pattern has shown up as a problem in the past for queries with complex expressions or lots of columns.